dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java

```diff
@@ -0,0 +1,209 @@
+   * @param conf
+   * @param alluxioEtcdClient
+   */
+  public EtcdMembershipManager(AlluxioConfiguration conf, AlluxioEtcdClient alluxioEtcdClient) {
+    mConf = conf;
+    mClusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    mRingPathPrefix = String.format(RING_PATH_FORMAT, mClusterName);
+    mAlluxioEtcdClient = alluxioEtcdClient;
+  }
+
+  @Override
+  public void join(WorkerInfo wkrAddr) throws IOException {
+    WorkerServiceEntity entity = new WorkerServiceEntity(wkrAddr.getAddress());
+    // 1) register to the ring
+    String pathOnRing = new StringBuffer()
+        .append(mRingPathPrefix)
+        .append(entity.getServiceEntityName()).toString();
+    byte[] ret = mAlluxioEtcdClient.getForPath(pathOnRing);
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    DataOutputStream dos = new DataOutputStream(baos);
+    entity.serialize(dos);
+    byte[] serializedEntity = baos.toByteArray();
+    // If there's existing entry, check if it's me.
+    if (ret != null) {
+      // It's not me, something is wrong.
+      if (!Arrays.equals(serializedEntity, ret)) {
+        throw new AlreadyExistsException(
+            "Some other member with same id registered on the ring, bail.");
+      }
+      // It's me, go ahead to start heartbeating.
+    } else {
+      // If haven't created myself onto the ring before, create now.
+      mAlluxioEtcdClient.createForPath(pathOnRing, Optional.of(serializedEntity));
+    }
+    // 2) start heartbeat
+    mAlluxioEtcdClient.mServiceDiscovery.registerAndStartSync(entity);
+  }
+
+  @Override
+  public List<WorkerInfo> getAllMembers() throws IOException {
+    List<WorkerServiceEntity> registeredWorkers = retrieveFullMembers();
+    return registeredWorkers.stream()
+        .map(e -> new WorkerInfo().setAddress(e.getWorkerNetAddress()))
+        .collect(Collectors.toList());
+  }
+
+  private List<WorkerServiceEntity> retrieveFullMembers() throws IOException {
+    List<WorkerServiceEntity> fullMembers = new ArrayList<>();
+    List<KeyValue> childrenKvs = mAlluxioEtcdClient.getChildren(mRingPathPrefix);
+    for (KeyValue kv : childrenKvs) {
+      try (ByteArrayInputStream bais =
+               new ByteArrayInputStream(kv.getValue().getBytes())) {
+        DataInputStream dis = new DataInputStream(bais);
+        WorkerServiceEntity entity = new WorkerServiceEntity();
+        entity.deserialize(dis);
+        fullMembers.add(entity);
+      } catch (IOException ex) {
+        // Ignore
+      }
+    }
+    return fullMembers;
+  }
+
+  private List<WorkerServiceEntity> retrieveLiveMembers() throws IOException {
+    List<WorkerServiceEntity> liveMembers = new ArrayList<>();
+    for (Map.Entry<String, ByteBuffer> entry : mAlluxioEtcdClient.mServiceDiscovery
+        .getAllLiveServices().entrySet()) {
+      try (ByteBufferInputStream bbis =
+               new ByteBufferInputStream(entry.getValue())) {
+        DataInputStream dis = new DataInputStream(bbis);
```

Comment:
Try with resources can be used with multiple resources
```suggestion
      try (ByteBufferInputStream bbis =
               new ByteBufferInputStream(entry.getValue());
          DataInputStream dis = new DataInputStream(bbis)) {
```

Reply:
done